### ISSUE TITLE
getGrants of specific role

### ISSUE THREAD

**Issue by sm2017** [onury/accesscontrol#62]
I have use case similar to https://github.com/onury/accesscontrol/issues/31

I don't  want expose all roles and action to client , just roles and actions that granted

In the following
```js
ac.grant('user').read('profile', ['*', '!id'])
   .grant('admin').extend('user');

{
    "admin": {
        "$extend": ["user"]
    },
    "user": {
        "profile": {
            "read:any": ["*", "!id"]
        }
    }
}
```

If the client's role is `user` , the admin role is exposed too

**Comment by onury** [comment]
This is not specific to AccessControl.js. You should never expose grants / policies to client side. 
_(Even if you only expose the policy of the required role and no other role; it's still bad practice.)_

You should handle the request and grant/deny access on server side. And if granted; filter data if needed (using `permission#filter()` if using this lib) and return the response. Client side should have no knowledge of your policies.

**Comment by sm2017** [comment]
@onury I know that I must grant/deny access on server side
As I told before I have use case similar to https://github.com/onury/accesscontrol/issues/31

I want to hide/show sidebar menu items based on grants, So I want to get all grants of logged in user and create new instance of AccessControl using the grants

**Comment by onury** [comment]
Currently there is no built-in way to get compiled resources and resource-attributes for a role, all at once. I'm marking this as a feature request.

For now you can use this function:
```js
const actions = ['create', 'read', 'update', 'delete'];
const possessions = ['Any', 'Own'];
function getCompiledGrantsOf(role) {
  const compGrants = {};
  let permission;
  ac.getResources().forEach(resource => {
    actions.forEach(act => {
      possessions.forEach(pos => {
        permission = ac.can(role)[act + pos](resource);
        if (permission.granted) {
          compGrants[resource] = compGrants[resource] || {};
          compGrants[resource][`${act}:${pos.toLowerCase()}`] = permission.attributes;
        }
      });
    });
  });
  return compGrants;
}
```

You can also:
- (**Recommended**) Implement **SSR** and render menu items on server side after AC handles it 
(e.g. SSR with React or Next.js has built-in support for this)

- Or create a new **resource** for such things (menu-items), setup your AC to grant it. Then fetch this from your endpoint as filtered data after AC handles it.
```js
// Let's say menu is a resource with attributes:
//    home, profile, videos, accounts
ac.grant('guest').read('menu', ['home'])
  .grant('user').read('menu', ['home', 'profile', 'videos'])
  .grant('admin').read('menu', ['*']);

// when checking and filtering, this is a bit different.
// instead of permission#filter we'll use Array#filter
const menuItems = ['home', 'profile', 'videos', 'accounts'];
const permission = ac.can(role).read('menu');
if (permission.granted) {
  const attrs = permission.attributes;
  if (attrs.indexOf('*') >= 0) return menuItems.concat();
  return menuItems.filter(m => attrs.indexOf(m) >= 0)
}

```
- Or pre-calculate the permissions required for these menu-items.
A simple example would be:

```js
// Let's say menu items are: 
//     Home, My Profile, Edit Profile, My Videos, All Videos

// on server-side
function getGrantedMenuItemsOf(role) {
    const menuItems = ['Home'];
    if (!role) return menuItems;

    const can = ac.can(role);
    if (can.readOwn('profile').granted) menuItems.push('My Profile');
    if (can.updateOwn('profile').granted) menuItems.push('Edit Profile');
    if (can.readAny('video').granted) {
        menuItems.push('My Videos');
        menuItems.push('All Videos');
    } else  if (can.readOwn('video').granted) {
        menuItems.push('My Videos');
    }

    return menuItems;
}

// Bind this to an API endpoint and fetch from client-side. e.g.:
getGrantedMenuItemsOf(req.user.role);
```


_Note: These are all quick thoughts. Since I have almost no knowledge of your application, possibly there might be better ways. This is a security concept so, you should carefully consider/improve/test, if you decide to implement._

**Comment by stalniy** [comment]
I don’t agree that you should not expose permissions. Your clients or users knows what they can do and they can’t do anyway (because you document this somewhere on app website). 

For example, on medium anonymous users can’t leave applause and it’s clear for everybody (because you just can’t do this on ui). 

So, what’s the difference whether you share permissions as JSON object or as human readable text or as UX?

If you don’t share permissions, your clients will need to duplicate or even hardcode permission logic. Later it will be very hard to change. 

I consider it to be safe to share permissions with client for currently logged in user. Some time ago I wrote an article about this:

https://medium.com/dailyjs/casl-and-cancan-permissions-sharing-between-ui-and-api-5f1fa8b4bec

**Comment by stalniy** [comment]
A good restful design requires to return 403 Forbidden status response for actions which are not allowed. So, anyway it’s quite easy to understand or get a list of permissions for a particular user credentials. 

**Side note**: I think the issue with permission sharing in this library exists because it relies on user roles and not on what user can do (e.g., create post, read post, leave comment). If you could share user actions instead of roles, it would not expose any internals of underlying permission logic

**Comment by onury** [comment]
Following the approach from this thread: the recommendation is **not** to ship grants/policy to the client at all — send a small **capability map** of booleans, computed server-side:

```js
const caps = {
  profile:  ac.tryCan(role).read('menu:profile').granted,
  videos:   ac.tryCan(role).read('menu:videos').granted,
  accounts: ac.tryCan(role).read('menu:accounts').granted,
};
res.json(caps); // client shows/hides menu items from these flags
```

(`tryCan()` never throws, so it's safe on the view path.) Or model the menu as a resource and return it as filtered data, or render server-side (SSR) — all in the [recipe](https://onury.io/accesscontrol/guides/recipes/). v3 intentionally still has no "dump a role's whole policy" getter, since the client shouldn't hold your policy.

Closing — happy to help if the capability-map approach doesn't fit your UI.

Sorry for the late reply!
